### ISSUE TITLE
Add examples to documentation for CudnnRNN

### ISSUE THREAD

**Issue by danielwatson6** [tensorflow/tensorflow#20273]
Have I written custom code: N/A
OS Platform and Distribution: N/A
TensorFlow installed from: N/A
TensorFlow version: 1.8.0
Bazel version: N/A
CUDA/cuDNN version: 9.0/7.0
GPU model and memory: NVIDIA GTX 1080 Ti
Exact command to reproduce: N/A

It would be very helpful to add usage examples on how to use `cudnn_rnn`, particularly on cross-compatibility between non-CUDA and CUDA-supporting devices. Users should be able to figure out how to save/restore weights to run their models with, say, `tf.nn.rnn_cell.LSTMCell` or `tf.contrib.cudnn_rnn.CudnnLSTM`.

There are classes that seem to do this (e.g. `tf.contrib.cudnn_rnn.CudnnLSTMSaveable`) but there are no easily accessible code examples showing how they should be used.

*A possible feature request*: would it be possible to have a high-level wrapper that makes this choice automatically based on the availability of a CUDA device, saving and restoring weights accordingly? This should be possible with `tf.test.is_gpu_available(cuda_only=True)`.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by melgor** [comment]
@danielwatson6 Could you point me if I'm correct?
1. While training, we can use just `CudnnLSTM`
2. When I want to freeze graph to pb, I should use `CudnnCompatibleLSTM` with `MultiRNNCell` and `stack_bidirectional_dynamic_rnn` (if  model was trained with such options)

Then, everything can be used in GPU and CPU.

If that pipeline is ok, it should be added to documentation?
Also question about `CudnnLSTMSaveable`, so what for this class was created?

**Comment by melgor** [comment]
I have manage to find a way to restore `CudnnLSTM` model on CPU. 
Here are jupyters with example: https://gist.github.com/melgor/41e7d9367410b71dfddc33db34cba85f

The main problem was `variable_scope` which need to be set in proper way.

**Comment by qlzh727** [comment]
"A possible feature request: would it be possible to have a high-level wrapper that makes this choice automatically based on the availability of a CUDA device, saving and restoring weights accordingly? This should be possible with tf.test.is_gpu_available(cuda_only=True)"

We are actually trying to unify the API for this exact direction. Basically the new API will figure out the best choice implementation for user based on the available hardware, and user don't have to specify whether its CuDNN or CPU implementation in the code. Currently we are still in the design stage, and will post more results when the APIs are finalized. The new API will probably be available as part of the change in Tensorflow 2.0.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by chenbiaolong** [comment]
@melgor  which version of tensorflow you use in your scrip?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 31 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 61 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 76 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 92 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 106 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by robieta** [comment]
Closing as RNNs will now automatically select an implementation based on whether or not they are running on a GPU.

**Comment by jchia** [comment]
@robieta Which specific classes implement the automatic implementation selection you talk about? Are you suggesting that `tensorflow.keras.layers.CuDNNGRU` is now no longer needed because of those classes that implement automatic implementation selection?

As of TF 1.14, `tensorflow.keras.layers.CuDNNGRU` still behaves differently than `tensorflow.keras.layers.GRU`. In my experiment, when I switch between these two in the graph, the output I get during training is different; they simply calculate different things (whether it due to different formulas or different initial weights I cannot tell). The former also takes a shorter time to train each epoch, so there is still a good reason to prefer it for training.

**Comment by robieta** [comment]
The automatic CuDNN selection is only for the 2.x RNNs.

**Comment by jchia** [comment]
@robieta Which 2.0 RNN classes have the automatic CuDNN selection?

**Comment by qlzh727** [comment]
@jchia, LSTM and GRU layer has auto cudnn selection in the 2.x API.

**Comment by Interfish** [comment]
@danielwatson6  
Hi, I know this is a 2018 old issue but I am kind of stuck at those `opaque kernel` in tf 1.14,  so may I talk more details with you?
The situation is , I got a trained model with cudnnBLSTM in tf 1.14. What I want is to convert this tf models to [MNN](https://github.com/alibaba/MNN) model format and then inference with that engine. I have a frozen graph of this model and I need to decode the weights and bias data from this frozen graph. 
What I found is that the size of the `opaque kernel` is not equal to the size I calculate of this BLSTM. Specifically, say I got a 3 layers BLSTM with input dim 143 and state size 143, so the total size of parameter should be
```
(143 * 143 + 143 * 143 + 143) * 4 * 2 + 2 * (286 * 143 + 143 * 143 + 143) * 4 * 2 = 1312168
```
But when I try to read the size of `opaque kernel`, I got 1315600.
So there is no suprise when I read your comments about

> The size and layout of the parameter buffers may change between
CUDA/CuDNN/GPU generations. Because of that, the opaque parameter variable
does not have a static shape and is not partitionable. 

So my question are, is there anyway I can decode weights and bias from `opaque kernel`? How does the 'opaque kernel' organized? Is there any workaround consider my situation?
Thanks!
